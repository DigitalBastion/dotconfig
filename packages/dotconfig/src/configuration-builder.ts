import type { IConfigurationBuilder, IConfigurationProvider, IConfigurationSource } from "./abstractions.js";
import { ConfigurationRoot } from "./configuration-root.js";
import { EnvironmentSource } from "./providers/environment.js";
import { MemoryConfigurationSource } from "./providers/memory.js";

export class ConfigurationBuilder implements IConfigurationBuilder {
    #sources: IConfigurationSource[] = [];
    #properties = new Map<string, unknown>();

    public get sources() {
        return this.#sources;
    }

    public get properties() {
        return this.#properties;
    }

    public add(source: IConfigurationSource): this {
        this.#sources.push(source);
        return this;
    }

    public addEnvironmentVariables(delimiter?: string): this {
        this.add(new EnvironmentSource({ delimiter }));
        return this;
    }

    public addMemoryCollection(collection: Map<string, string | null>): this {
        this.add(new MemoryConfigurationSource(collection));
        return this;
    }

    public async build() {
        const providerPromises: Array<Promise<IConfigurationProvider>> = this.#sources.map(source => source.build(this));
        const providers = await Promise.all(providerPromises);
        return new ConfigurationRoot(providers);
    }
}