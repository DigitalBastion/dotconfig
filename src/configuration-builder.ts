import type { ConfigurationProvider } from "./configuration-provider.js";
import { Configuration } from "./configuration.js";
import { DELIMITER } from "./constants.js";
import {
  EnvironmentConfigurationProvider,
  type EnvironmentConfigurationProviderOptions,
} from "./providers/environment.js";
import { InMemoryConfigurationProvider } from "./providers/in-memory.js";
import { JsonFileConfigurationProvider, type JsonFileConfigurationProviderOptions } from "./providers/json-file.js";
import { flatten } from "./utils/flatten.js";

export interface ConfigBuilderOptions {
  logger?: Console;
}

export class ConfigurationBuilder {
  constructor(protected options: ConfigBuilderOptions = {}) {}

  protected providers: ConfigurationProvider[] = [];

  public addEnvironmentVariables(options: EnvironmentConfigurationProviderOptions = {}): this {
    const provider = new EnvironmentConfigurationProvider(options);
    this.providers.push(provider);
    return this;
  }

  public addJsonFile(options: JsonFileConfigurationProviderOptions): this {
    const provider = new JsonFileConfigurationProvider(options);
    this.providers.push(provider);
    return this;
  }

  public addInMemory(data: Record<string, unknown>): this {
    const provider = new InMemoryConfigurationProvider(data);
    this.providers.push(provider);
    return this;
  }

  public addProvider(provider: ConfigurationProvider): this {
    this.providers.push(provider);
    return this;
  }

  public async build(): Promise<Configuration> {
    const flatConfig = new Map<string, string>();

    for (const provider of this.providers) {
      const data = await provider.load();

      const flattenData = flatten(data, {
        delimiter: DELIMITER,
      });

      for (const [key, value] of flattenData) {
        if (value == null) {
          continue;
        }

        flatConfig.set(key, value.toString());
      }
    }
    const configuration = new Configuration(flatConfig);

    return configuration;
  }
}
