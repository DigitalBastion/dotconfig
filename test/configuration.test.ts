import { expect, test } from "bun:test";
import { ConfigurationBuilder } from "../src/configuration-builder.js";

test("It should produce configuration instance.", async () => {
  const builder = new ConfigurationBuilder().addInMemory({
    Logging: {
      Enabled: true,
      LogLevel: "Info",
    },
    Database: {
      ConnectionString: "test_connection",
      MaxConnections: 100,
    },
    App: {
      Settings: {
        Version: "1.0.0",
        Name: "TestApp",
      },
    },
  });

  const configuration = await builder.build();

  expect(configuration.toObject()).toMatchSnapshot();
});
