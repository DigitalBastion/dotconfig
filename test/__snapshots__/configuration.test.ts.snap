// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`It should produce configuration instance. 1`] = `
{
  "App": {
    "Settings": {
      "Name": "TestApp",
      "Version": "1.0.0",
    },
  },
  "Database": {
    "ConnectionString": "test_connection",
    "MaxConnections": "100",
  },
  "Logging": {
    "Enabled": "true",
    "LogLevel": "Info",
  },
}
`;
